README.md
# go-challenge

Task - Retrieve data from URLs, merge, filter duplicates and sort in 500ms.


The rationale behind the various decisions taken while developing the solution are explained below. Please read the comments in code for better understanding. If you wish to run the application in docker, a dockerfile is included. The application exposes an endpoint called /numbers and listens on port 8000.

## Context package
The first thing that popped into my head when I saw "500ms" was go's context package.
The other alternatives which I considered were: 
* Using a http client with a timeout
* Wait on a channel inside the select statement using time.After. 

The rationale behind using context is cancellation propagation. In the aforementioned alternatives, there is no way to signal to the various goroutines to stop their task and return immediately after a certain time. But context provides a clean, easy and efficient way of doing this. This makes http clients and servers more efficient and less wasteful. For e.g. if a client that requested some resource from the server suddenly drops off, context helps us to cancel all the functions and goroutines that were running to satisfy the client's request. Context achieves cancellation propagation by maintaining a tree structure of "context dependencies". Also, cancellation is transitive i.e. if a particular node in the context tree is cancelled, all the child nodes are automatically cancelled.

## Timeout
The next question to be tackled was - "How long do we let the goroutines that fetch data from valid URLs execute?". The requirements specified that the /numbers endpoint needs to return the results within 500ms, but no timeout was specified for the individual URLs that were passed as query params. Since results had to be returned within 500ms, the timeout for each goroutine was set to 450ms after which it would be cancelled and remaining 50ms would be utilized to sort the accumulated results. However for a real-world application we would have to consider the following criteria before deciding the timeout:

* The size of the result returned by each of the URLs in the query parameter.
* How sorted or unsorted are each of the results, returned by the URLs?
* Performance of go's sort package.

## Buffered vs unbuffered channels
Once the timeout for the goroutines was decided, the next question to answer was "Whether to use buffered or unbuffered channel?" Buffered channel was the obvious choice because we didn't want the various goroutines that were spawned to fetch results from each URL to be blocked by each other, waiting for the channel to be free. Also, we didn't want the goroutines to be blocked by our receiver (the main goroutine in our case).

## Filtering duplicates and sorting
The next problem to solve was "When to filter duplicates and sort the results returned by the various URLs?". 

* Filter duplicates and sort after the results from all the URLs have been accumulated or do it as and when the result from each URL are fetched?
* Use the built-in sort package or implement own sorting technique?
* Can the performance be improved by using data structures such as heaps?

The chosen solution to this problem was to remove duplicates as and when the URLs return results by maintaining a hashtable (since go does not have a native implementation of set) of all the encountered values. When new results are received, we check if the values present in the result are already present in the hashtable. If not, those values are added both to the hashtable and final result array. Hashtable was chosen as it provides lookups in O(1) and does not allow duplicates. Once duplicates are filtered across the results returned by all the URLs, we sort the final result array before sending it over the wire. Sorting is done once 450ms have elapsed because sorting the results of individual URLs is wasteful and repetitive. Heap wasn't considered because we would have to iterate over the results to push it to the heap and iterate once more once we have results from all the URLs to pop individual elements from the heap. The built-in sort package was chosen after looking at the benchmarks published [here](https://stackimpact.com/blog/practical-golang-benchmarks/#sorting).

## What to do with errors?
For now errors are just being logged. The errors from the various URLs are collected from the goroutines using an error channel and logged in the main goroutine. Alternatively, this could be sent to the client.

changes.md
### Handling a large number of URLs

#### Counting semaphore
I started by implementing a counting semaphore wherein I would spawn all the required goroutines and only x number of goroutines would get a place on the semaphore channel.

```go
sem := make(chan struct{}, min(maxConnections, len(urls)))

func fetch(){
    sem <- struct{}{}
    defer func() { <-sem }()
}
```

If the number of URLs = 100000, we would be spawning 100000 goroutines of which only x, say 500, would get a place on the semaphore channel and actually execute the fetch logic. The rest of the goroutines would be busy waiting for a place on the semaphore channel.

#### Worker pool
After some careful deliberation and researching on various go forums and blog posts, the suggested method of solving this problem was to spawn x number of worker goroutines and perform the fetch operation. This ensures that no goroutines waste time waiting and that we do not hit the socket limit or the file descriptor limits. During benchmarking on my laptop, the application successfully handled 100000 URLs and will probably handle more but might take significant time.

PS. I have increased both context timeouts and individual request timeouts to accommodate for the time taken to process a large number of URLs and larger datasets. As a result, some tests in the test table might fail. 

Also, benchmarks have been added to the test file. 

### De-duplication and sorting

Filtering duplicates across results returned by the URLs would be a trade-off between time and memory. The time efficient way of doing it is by using a Set(map in our case) which holds distinct elements. The other way of doing this (space efficient) would be by linearly comparing all elements and filtering duplicates.


Before discussing the sorting techniques, the question about when to sort should be handled.
* Continuous de-duplication and continuous sorting - Every time a URL responds, we filter duplicates, merge the data with an "accumulator" slice and sort it. This would be less efficient as we would sort the data repeatedly and sorting is computationally expensive. But an advantage of this approach is that any given time, the results would be sorted and ready to be sent over the wire.

* Continuous de-duplication and sorting once - In this approach, we filter duplicates, merge the data with an "accumulator" slice every time a URL responds.  Once all the URLs have responded, we sort the accumulated data. This would be significantly efficient since we only sort once after the results from all the URLs have been accumulated. On the flip side, the results wouldn't be sorted to be sent over the wire at any given time.

The following approaches were considered to tackle the problem of sorting
* built-in sort package - Uses a combination of quicksort, shellsort and insertionsort. Generalized algorithm hence performs well in most general cases. O(nlogn)
* radix sort - Works best on integers. Sorts by comparing each digit of a number. Might perform badly if the number of digits in a number is huge. This is because radix sort performs a "sorting pass" which uses counting sort for each digit in the numbers of the data set. O(kn) where k is the number of digits in the largest number.
* heapsort - Gets instant access to the largest/smallest element in O(1). Was curious how a heap would perform in our case. 
* b-trees - Automatic sorting and de-duplicating of elements.

PS. For b-trees and radix sort I used external packages 
* radix sort - https://github.com/shawnsmithdev/zermelo - This package claims the following - "You will generally only want to use zermelo if you won't mind the extra memory used for buffers and your application frequently sorts slices of supported types with at least 256 elements (128 for 32-bit types). The larger the slices you are sorting, the more benefit you will gain by using zermelo instead of the standard library's in-place comparison sort." 
* b-trees - https://github.com/google/btree 

### Benchmarks

The benchmarks for the sorting algorithms can be found at https://github.com/karthikraobr/go-sorting-bench. To mimic the problem at hand, a slice of 1 million int values is being split into 10 - 100,000 slices.

* filterAndSortOnceDefault - Filters duplicates as and when URLs respond and sorts once all URLs have responded, using the built-in sort package.
* filterAndSortOnceZermelo - Same as above but uses zermelo radix sort package.
* filterAndContinuousSortDefault - Filters duplicates as and when URLs respond and sort them on a per-URL basis. Sort uses built-in sort package.
* filterAndContinuousSortZermelo - Same as above but uses zermelo radix sort package.
* heapsort - Filter duplicates and push into a heap. Then pop one element at a time.
* btree - Push each element of the URL responses into a tree. De-duplication and sorting is taken care by the implementation.

| Name        | Count           | Time taken  | Memory
| ------------- |:-------------:| -----:|-----:|
|BenchmarkEverything/filterAndSortOnceDefault-8                    |5     |296403520 ns/op    |53860502 B/op      | 19206 allocs/op|
|BenchmarkEverything/filterAndSortOnceZermelo-8                   |10     |162799810 ns/op    |58916528 B/op      | 19170 allocs/op|
|BenchmarkEverything/filterAndContinuousSortDefault-8              |2     |749996950 ns/op    |53849292 B/op       |19186 allocs/op|
|BenchmarkEverything/filterAndContinuousSortZermelo-8              |5     |262995680 ns/op    |85795990 B/op      | 19044 allocs/op|
|BenchmarkEverything/heap-8                                       |2     |552999600 ns/op    |92791836 B/op     |1283001 allocs/op|
|BenchmarkEverything/btree-8                                       |1    |1335997200 ns/op    |52337000 B/op     |1178894 allocs/op|


server_test.go
package main

import (
	"bytes"
	"encoding/json"
	"math/rand"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

const (
	localhost      = "localhost:8000"
	invalidRequest = "InvalidRequest"
	invalidURL     = "InvalidURL"
	randomURL      = "RandomURL"
	forbiddenTest  = "403Test"
)

func Test_numberHandler(t *testing.T) {
	actual := []int{1, 1, 2, 3, 5, 8, 13, 21}
	expected := []int{1, 2, 3, 5, 8, 13, 21}
	tt := []struct {
		name     string
		handler  func(http.ResponseWriter, *http.Request)
		expected result
	}{
		{name: "Simple", handler: simpleHandler(actual), expected: result{Numbers: expected}},
		{name: "NoParam", handler: nil, expected: result{Numbers: []int{}}},
		{name: "InvalidURL", handler: nil, expected: result{Numbers: []int{}}},
		{name: "InvalidRequest", handler: nil, expected: result{Numbers: []int{}}},
		{name: "RandomURL", handler: nil, expected: result{Numbers: []int{}}},
		{name: "SimpleError", handler: errHandler(), expected: result{Numbers: []int{}}},
		{name: "SimpleTimeOut", handler: timeOutHandler(actual), expected: result{Numbers: []int{}}},
		{name: "JustInTime", handler: justInTimeHandler(actual), expected: result{Numbers: expected}},
		{name: "ErrorAfterTime", handler: errAfterTimeHandler(), expected: result{Numbers: []int{}}},
		{name: forbiddenTest, handler: nil, expected: result{Numbers: []int{}}},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			var req *http.Request
			var err error
			if tc.handler != nil {
				ts := httptest.NewServer(http.HandlerFunc(tc.handler))
				defer ts.Close()
				req, err = http.NewRequest(http.MethodGet, localhost+"?u="+ts.URL, nil)
				if err != nil {
					t.Fatalf("could not create request: %v", err)
				}
			} else {
				if tc.name == invalidRequest {
					req, err = http.NewRequest(http.MethodGet, localhost+"?u=hello", nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				} else if tc.name == invalidURL {
					req, err = http.NewRequest(http.MethodGet, localhost+"?u=http://\\www.google.com//", nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				} else if tc.name == randomURL {
					req, err = http.NewRequest(http.MethodGet, localhost+"?u=http://www.google.com", nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				} else if tc.name == forbiddenTest {
					req, err = http.NewRequest(http.MethodPost, localhost+"?u=http://www.google.com", nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				} else {
					req, err = http.NewRequest(http.MethodGet, localhost, nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				}
			}
			rec := httptest.NewRecorder()
			numbersHandler(rec, req)
			res := rec.Result()
			defer res.Body.Close()
			if tc.name != forbiddenTest {
				if res.StatusCode != http.StatusOK {
					t.Fatalf("expected status OK; got %v", res.Status)
				}
				var num result
				if err = json.NewDecoder(res.Body).Decode(&num); err != nil {
					t.Fatalf("could not decode response: %v", err)
				}
				if !num.equals(tc.expected) {
					t.Errorf("expected %v but got %v", tc.expected, num)
				}
			} else if res.StatusCode != http.StatusForbidden {
				t.Fatalf("expected status forbidden; got %v", res.Status)
			}
		})
	}
}

func simpleHandler(numbers []int) func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		json.NewEncoder(w).Encode(map[string]interface{}{"numbers": numbers})
	}
}

func errHandler() func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "service unavailable", http.StatusServiceUnavailable)
	}
}

func timeOutHandler(numbers []int) func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(time.Duration(451) * time.Millisecond)
		json.NewEncoder(w).Encode(map[string]interface{}{"numbers": numbers})
	}
}

func justInTimeHandler(numbers []int) func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(time.Duration(440) * time.Millisecond)
		json.NewEncoder(w).Encode(map[string]interface{}{"numbers": numbers})
	}
}

func errAfterTimeHandler() func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(time.Duration(300) * time.Millisecond)
		http.Error(w, "service unavailable", http.StatusServiceUnavailable)
	}
}
func (a *result) equals(b result) bool {
	if a.Numbers == nil && b.Numbers == nil {
		return true
	}

	if a.Numbers == nil || b.Numbers == nil {
		return false
	}

	if len(a.Numbers) != len(b.Numbers) {
		return false
	}

	for i := range a.Numbers {
		if a.Numbers[i] != b.Numbers[i] {
			return false
		}
	}
	return true
}

func BenchmarkNumbersHandler(b *testing.B) {
	ts := httptest.NewServer(http.HandlerFunc(simpleHandler([]int{1, 1, 2, 3, 5, 8, 13, 21})))
	var buf bytes.Buffer
	for i := 0; i < 100000; i++ {
		buf.WriteString("&u=")
		buf.WriteString(ts.URL)
	}
	req, _ := http.NewRequest(http.MethodGet, localhost+"?u="+ts.URL+buf.String(), nil)
	tsLoad := httptest.NewServer(http.HandlerFunc(loadHandler()))

	reqLoad, _ := http.NewRequest(http.MethodGet, localhost+"?u="+tsLoad.URL+"&u="+tsLoad.URL+"&u="+tsLoad.URL+"&u="+tsLoad.URL+"&u="+tsLoad.URL+"&u="+tsLoad.URL+"&u="+tsLoad.URL+"&u="+tsLoad.URL+"&u="+tsLoad.URL+"&u="+tsLoad.URL, nil)
	tt := []struct {
		name string
		req  *http.Request
	}{
		{name: "tonsOfUrls", req: req},
		{name: "tonsOfNumbers", req: reqLoad},
	}

	for _, tc := range tt {
		b.Run(tc.name, func(b *testing.B) {
			for i := 0; i < b.N; i++ {
				rec := httptest.NewRecorder()
				numbersHandler(rec, tc.req)
				res := rec.Result()
				defer res.Body.Close()
			}
		})
	}
}

func loadHandler() func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		json.NewEncoder(w).Encode(map[string]interface{}{"numbers": foo()})
	}
}

func foo() []int {
	hold := make([]int, 1000000)
	for i := 0; i < 1000000; i++ {
		a := rand.Intn(100000000)
		hold = append(hold, a)
	}
	return hold
}


numbers_test.go
package main

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

const (
	localhost      = "localhost:8000"
	invalidRequest = "InvalidRequest"
	invalidURL     = "InvalidURL"
	randomURL      = "RandomURL"
	forbiddenTest  = "403Test"
)

func Test_numberHandler(t *testing.T) {
	actual := []int{1, 1, 2, 3, 5, 8, 13, 21}
	expected := []int{1, 2, 3, 5, 8, 13, 21}
	tt := []struct {
		name     string
		handler  func(http.ResponseWriter, *http.Request)
		expected result
	}{
		{name: "Simple", handler: simpleHandler(actual), expected: result{Numbers: expected}},
		{name: "NoParam", handler: nil, expected: result{Numbers: []int{}}},
		{name: "InvalidURL", handler: nil, expected: result{Numbers: []int{}}},
		{name: "InvalidRequest", handler: nil, expected: result{Numbers: []int{}}},
		{name: "RandomURL", handler: nil, expected: result{Numbers: []int{}}},
		{name: "SimpleError", handler: errHandler(), expected: result{Numbers: []int{}}},
		{name: "SimpleTimeOut", handler: timeOutHandler(actual), expected: result{Numbers: []int{}}},
		{name: "JustInTime", handler: justInTimeHandler(actual), expected: result{Numbers: expected}},
		{name: "ErrorAfterTime", handler: errAfterTimeHandler(), expected: result{Numbers: []int{}}},
		{name: forbiddenTest, handler: nil, expected: result{Numbers: []int{}}},
	}

	for _, tc := range tt {
		t.Run(tc.name, func(t *testing.T) {
			var req *http.Request
			var err error
			if tc.handler != nil {
				ts := httptest.NewServer(http.HandlerFunc(tc.handler))
				defer ts.Close()
				req, err = http.NewRequest(http.MethodGet, localhost+"?u="+ts.URL, nil)
				if err != nil {
					t.Fatalf("could not create request: %v", err)
				}
			} else {
				if tc.name == invalidRequest {
					req, err = http.NewRequest(http.MethodGet, localhost+"?u=hello", nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				} else if tc.name == invalidURL {
					req, err = http.NewRequest(http.MethodGet, localhost+"?u=http://\\www.google.com//", nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				} else if tc.name == randomURL {
					req, err = http.NewRequest(http.MethodGet, localhost+"?u=http://www.google.com", nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				} else if tc.name == forbiddenTest {
					req, err = http.NewRequest(http.MethodPost, localhost+"?u=http://www.google.com", nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				} else {
					req, err = http.NewRequest(http.MethodGet, localhost, nil)
					if err != nil {
						t.Fatalf("could not create request: %v", err)
					}
				}
			}
			rec := httptest.NewRecorder()
			numberHandler(rec, req)
			res := rec.Result()
			defer res.Body.Close()
			if tc.name != forbiddenTest {
				if res.StatusCode != http.StatusOK {
					t.Fatalf("expected status OK; got %v", res.Status)
				}
				var num result
				if err = json.NewDecoder(res.Body).Decode(&num); err != nil {
					t.Fatalf("could not decode response: %v", err)
				}
				if !num.equals(tc.expected) {
					t.Errorf("expected %v but got %v", tc.expected, num)
				}
			} else if res.StatusCode != http.StatusForbidden {
				t.Fatalf("expected status forbidden; got %v", res.Status)
			}
		})
	}
}

func simpleHandler(numbers []int) func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		json.NewEncoder(w).Encode(map[string]interface{}{"numbers": numbers})
	}
}

func errHandler() func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		http.Error(w, "service unavailable", http.StatusServiceUnavailable)
	}
}

func timeOutHandler(numbers []int) func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(time.Duration(451) * time.Millisecond)
		json.NewEncoder(w).Encode(map[string]interface{}{"numbers": numbers})
	}
}

func justInTimeHandler(numbers []int) func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(time.Duration(440) * time.Millisecond)
		json.NewEncoder(w).Encode(map[string]interface{}{"numbers": numbers})
	}
}

func errAfterTimeHandler() func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(time.Duration(300) * time.Millisecond)
		http.Error(w, "service unavailable", http.StatusServiceUnavailable)
	}
}
func (a *result) equals(b result) bool {
	if a.Numbers == nil && b.Numbers == nil {
		return true
	}

	if a.Numbers == nil || b.Numbers == nil {
		return false
	}

	if len(a.Numbers) != len(b.Numbers) {
		return false
	}

	for i := range a.Numbers {
		if a.Numbers[i] != b.Numbers[i] {
			return false
		}
	}
	return true
}


server.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"net/http"
	_ "net/http/pprof"
	"sort"
	"time"
)

const (
	endpoint = "/numbers"
	// The below 3 values should reside as environment variables for flexibility
	// Max number of simultaneous workers
	maxConnections = 200
	// Timeout for requests. This is high in case the result of each URL contains millions of digits
	individualTimeout = 50000
	// Timeout for context. This is high in case we have to process large number of URLs
	timeout = 50000
)

//Type which represents the response of the given URLs as well as our response
type result struct {
	Numbers []int `json:"numbers"`
}

type payload struct {
	res chan result
	err chan error
}

func main() {
	listenAddr := flag.String("http.addr", ":8000", "http listen address")
	flag.Parse()
	http.HandleFunc(endpoint, numbersHandler)
	log.Fatal(http.ListenAndServe(*listenAddr, nil))
}

func numbersHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		w.WriteHeader(http.StatusForbidden)
		w.Write([]byte("403 - Method not supported!"))
		return
	}
	ctx := r.Context()
	ctx, cancel := context.WithTimeout(ctx, timeout*time.Millisecond)
	defer cancel()
	u := r.URL
	q := u.Query()
	params := q["u"]
	if len(params) == 0 {
		json.NewEncoder(w).Encode(result{Numbers: []int{}})
	} else {
		// Create the http transport for reuse
		t := &http.Transport{
			Proxy:               http.ProxyFromEnvironment,
			MaxIdleConnsPerHost: maxConnections,
			// Timeout for individual requests
			ResponseHeaderTimeout: individualTimeout * time.Millisecond,
		}
		res := make(chan result, maxConnections)
		err := make(chan error, maxConnections)
		p := payload{res: res, err: err}
		// Spawn go routines for worker to consume
		go fetchAll(ctx, t, params, &p)
		// Consumer to consume from channels
		json.NewEncoder(w).Encode(result{Numbers: consume(ctx, len(params), &p)})
	}
}

// Spawns worker goroutines and generate work
func fetchAll(ctx context.Context, t *http.Transport, urls []string, p *payload) {
	c := make(chan string)
	// Spin up workers. Only 200 workers will be concurrently fetching from URLs.
	// This will ensure we do not run out of sockets or hit file descriptor limits
	for i := 0; i < maxConnections; i++ {
		go doWork(ctx, t, c, p)
	}
	// Queue up work by putting URLs in a queue. The doWork goroutine will consume this channel.
	for _, u := range urls {
		c <- u
	}
	// Closing channel to indicate to doWork that we have processed all URLs
	close(c)
}

func doWork(ctx context.Context, t *http.Transport, u chan string, p *payload) {
	// Consume URLs until the channel is closed
	for {
		url, ok := <-u
		//Channel is closed signaling all URLs have been processed
		if !ok {
			return
		}
		fetch(ctx, t, url, p)
	}
}

func fetch(ctx context.Context, t *http.Transport, u string, p *payload) {
	var number result
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		p.err <- fmt.Errorf("%s returned an error while creating a request- %v", u, err)
		return
	}
	req = req.WithContext(ctx)
	res, err := t.RoundTrip(req)
	if err != nil {
		p.err <- fmt.Errorf("%s returned an error while performing a request  - %v", u, err)
		return
	}
	// Close body so that sockets can be reused.
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		p.err <- fmt.Errorf("%s server returned an error - %v", u, res.Status)
		return
	}
	if err := json.NewDecoder(res.Body).Decode(&number); err != nil {
		p.err <- fmt.Errorf("%s decoding error - %v", u, err)
		return
	}
	//log.Println("success")
	p.res <- number
}

// Consumer to drain result and error channel. Also handles context timeouts.
func consume(ctx context.Context, count int, p *payload) []int {
	accumulator := make([]int, 0)
	visited := make(map[int]struct{})
	for i := 0; i < count; i++ {
		select {
		case res := <-p.res:
			for _, val := range res.Numbers {
				if _, ok := visited[val]; !ok {
					accumulator = append(accumulator, val)
					visited[val] = struct{}{}
				}
			}
		case err := <-p.err:
			log.Println(err)
		case <-ctx.Done():
			log.Println(ctx.Err())
			sort.Ints(accumulator)
			return accumulator
		}
	}
	sort.Ints(accumulator)
	return accumulator
}


numbers.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"sort"
	"time"
)

const (
	//Represents the context timeout. Can be tuned depending on requirements and benchmarks.
	timeout  = 450
	endpoint = "/numbers"
)

//Type which represents the response of the given URLs as well as our response
type result struct {
	Numbers []int `json:"numbers"`
}

func main() {
	listenAddr := flag.String("http.addr", ":8000", "http listen address")
	flag.Parse()
	http.HandleFunc(endpoint, numberHandler)
	log.Fatal(http.ListenAndServe(*listenAddr, nil))
}

//Http handler that handles the /numbers endpoint
func numberHandler(w http.ResponseWriter, r *http.Request) {
	//We support only "GET" method.
	if r.Method != http.MethodGet {
		w.WriteHeader(http.StatusForbidden)
		w.Write([]byte("403 - Method not supported!"))
		return
	}
	//Grab the context from the request. Incase a client drops we can cancel all the "tasks"
	//that were spawned for the client.
	ctx := r.Context()
	//Since the response needs to be sent on the wire before 500ms, we set the context timeout to 450ms
	//and use the remaining 50ms to sort the array.
	ctx, cancel := context.WithTimeout(ctx, timeout*time.Millisecond)
	//The cancel function signals the gc to collect resources allocated for context timers.
	defer cancel()
	u := r.URL
	q := u.Query()
	//Get all the query parameters with key "u".
	params := q["u"]
	//If there are no "u" query parameters, simply return an empty array.
	if len(params) == 0 {
		json.NewEncoder(w).Encode(result{Numbers: []int{}})
	} else {
		numbers := validateAndFetch(ctx, params)
		json.NewEncoder(w).Encode(result{Numbers: numbers})
	}
}

func validateAndFetch(ctx context.Context, urls []string) []int {
	//Accumulator holds non duplicate values returned by all the URLs.
	accumulator := make([]int, 0)
	//Map to eliminate duplicates across responses. We actually need a set,
	//but since go doesn't provide a set implementation, we use an empty struct so as
	//to not waste precious memory :) https://play.golang.org/p/ea_19tva-0T
	visited := make(map[int]struct{})
	//Buffered channel - So as to not block goroutines.
	//In the case of unbuffered channel, the sender is blocked when the channel is full and receiver
	//is blocked when the channel is empty. If the receiver is busy with other tasks and is taking a
	//long time to receive on the channel, all the senders are blocked. With a buffered channel, sends
	//are blocked when the channel has reached its maximum capacity and receives are blocked when the
	//channel is empty. Hence the sender has a "buffer" to send on and is not blocked by the "slow" receiver.
	//In our case the receiver (main goroutine) might be filtering duplicates from a slice, while other goroutines
	//send on the buffered channel.
	ch := make(chan result, len(urls))
	//An error channel to hold the various errors that might occur while performing the request. For now we just
	//log it. We can define retry policies such as exponential backoff based on various error types.
	er := make(chan error, len(urls))
	//Counter to keep track of number of valid URLs in the request.
	counter := 0
	//Check if all URLs in the request are valid and if so spawn a goroutine to fetch data.
	for _, u := range urls {
		_, err := url.Parse(u)
		if err != nil {
			log.Printf("%s returned an error- %v", u, err)
			continue
		}
		counter++
		//Spawn a goroutine for each valid URL.
		go fetch(ctx, u, ch, er)
	}

	//Loop to drain channels, filter out duplicates and check for timeout. Each goroutine either fills the
	//result channel or the error channel.
	for i := 0; i < counter; i++ {
		select {
		case res := <-ch:
			for _, val := range res.Numbers {
				//Eliminate duplicates. The rationale behind eliminating duplicates on a per-goroutine basis
				//as soon we receive on the channel rather than accumulating results from all the valid URLs
				//or after 450ms has elapsed is that - if no other goroutine has filled the channel, rather
				//than wasting precious processor clock on waiting, we may as well remove duplicates from
				//the array during that time.
				if _, ok := visited[val]; !ok {
					accumulator = append(accumulator, val)
					visited[val] = struct{}{}
				}
			}
		case err := <-er:
			log.Println(err)
			//After 450ms have elapsed, the context is finished. Done returns a closed channel that signals that
			//the context was cancelled, which in our case that is a timeout.
		case <-ctx.Done():
			log.Println(ctx.Err())
			//Sort and return as soon as the context is finished rather than waiting for other goroutines to be cancelled.
			sort.Ints(accumulator)
			return accumulator
		}
	}
	//Sort and return if all URLs respond within 450ms.
	sort.Ints(accumulator)
	return accumulator
}

func fetch(ctx context.Context, u string, c chan<- result, e chan<- error) {
	var number result
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		e <- fmt.Errorf("%s returned an error while creating a request- %v", u, err)
		return
	}
	//Perform a request with a context to enable cancellation propagation after 450ms has elapsed.
	//As soon 450ms is elapsed the parent conext signals all the goroutines to abandon their work and return.
	req = req.WithContext(ctx)
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		e <- fmt.Errorf("%s returned an error while performing a request  - %v", u, err)
		return
	}
	//Close response body as soon as function returns to prevent resource lekage.
	defer res.Body.Close()

	//If not 200 log the error.
	if res.StatusCode != http.StatusOK {
		e <- fmt.Errorf("%s server returned an error - %v", u, res.Status)
		return
	}
	if err := json.NewDecoder(res.Body).Decode(&number); err != nil {
		e <- fmt.Errorf("%s decoding error - %v", u, err)
		return
	}
	c <- number
}
